### PR TITLE
Generalise mux over the set of protocol ids

### PR DESCRIPTION
**Pull request by dcoutts**
Here for comment initially. Needs to be rebased now that #311 was merged. Should be easy to fix up.

We will need to instantiate the mux with different sets of mini-protocols for the node-to-node vs the node-to-client protocols.

The approach we take is a simple trick using custom bounded enumerations. This makes use of the Ix type, arrays and functions covering the whole range of the enumeration.

This also eliminates some of the Data.Map lookup error conditions.

Also update the Pipe and Socket demos to use this.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/NodeToClient.hs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+--
+module Ouroboros.Network.NodeToClient (
+  NodeToClientProtocols(..)
+  ) where
+
+import Ouroboros.Network.Mux.Types (ProtocolEnum(..))
+
+
+-- | An index type used with the mux to enumerate all the mini-protocols that
+-- make up the overall node-to-node protocol.
```
Comment (karknu) [inline ouroboros-network/src/Ouroboros/Network/NodeToClient.hs]: Should it be node-to-client here?

File: ouroboros-network/src/Ouroboros/Network/Pipe.hs [hunk 2]
```diff
@@ -88,11 +87,22 @@ readPipe ctx = do
...
 startPipe mpds (r, w) = do
     let ctx = PipeCtx r w
     setupMux mpds ctx
 
+-- | The enumeration of all mini-protocols in our demo protocol.
+data DemoProtocols = ChainSync
+  deriving (Eq, Ord, Enum, Bounded)
+
+instance Mx.ProtocolEnum DemoProtocols where
+  fromProtocolEnum ChainSync = 0
```
Comment (karknu) [inline ouroboros-network/src/Ouroboros/Network/Pipe.hs]: Must be 2.

File: ouroboros-network/src/Ouroboros/Network/Socket.hs [hunk 3]
```diff
@@ -129,6 +129,17 @@ hexDump :: BL.ByteString -> String -> IO ()
 hexDump buf out | BL.empty == buf = say out
 hexDump buf out = hexDump (BL.tail buf) (out ++ printf "0x%02x " (BL.head buf))
 
+-- | The enumeration of all mini-protocols in our demo protocol.
+data DemoProtocols = ChainSync
+  deriving (Eq, Ord, Enum, Bounded)
+
+instance Mx.ProtocolEnum DemoProtocols where
+  fromProtocolEnum ChainSync = 0
```
Comment (karknu) [inline ouroboros-network/src/Ouroboros/Network/Socket.hs]: As in Pipe.hs, this must be 2. 0 is MuxControl.

File: ouroboros-network/src/Ouroboros/Network/NodeToClient.hs (hunk 1 above)
Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/NodeToClient.hs]: Well spotted.

File: ouroboros-network/src/Ouroboros/Network/Pipe.hs (hunk 2 above)
Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/Pipe.hs]: Indeed.

File: ouroboros-network/src/Ouroboros/Network/Socket.hs (hunk 3 above)
Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/Socket.hs]: Yep, thanks!

**Comment by dcoutts** [comment]
All comments should be addressed. Rebased and tests fixed up. Should be ready to go.

**Comment by karknu** [comment]
LGTM.

**Pull request merged**